### ISSUE TITLE
[Feature Request] for Pull Request, Resolve Conflicts via WEB

### ISSUE THREAD

**Issue by 6543** [go-gitea/gitea#9014]
this is part of #5158 summary ... and could be an addition of #9013

**Comment by 6543** [comment]
https://stackoverflow.com/a/6283843
this solution prints out the merge wich can be also displaied in our web editor to resolve this ... ?

**Comment by UBowt571** [comment]
I'm new to gitea, but it looks like this button (under GitHub)
![image](https://user-images.githubusercontent.com/34211283/71891158-4565e980-3146-11ea-9178-1168b91260f1.png)  
has no equivalent under gitea and I must solve PR manually ?
Thanks

**Comment by 6543** [comment]
@UBowt571 until now ... yes :/

**Comment by MarcelHillmann** [comment]
+1

**Comment by UBowt571** [comment]
Is there someone already working on it ?
I think about trying to implement it (I'm a noob and this would be my very first fork) but I don't want to do the same work as someone else did

**Comment by guillep2k** [comment]
@UBowt571 All contributions are welcome! 😄

**Comment by brymut** [comment]
I'll try to have a go at this other the next week.

**Comment by nghtstr** [comment]
+1

**Comment by maskym** [comment]
@richmahn Ho, go ahead, I surrendered (normal as a French guy lol) and switched to self-hosted gitlab instead

**Comment by richmahn** [comment]
@maskym Haha. Loved the French joke. I've been working on Gitea and with Go since 2016, even the one who added the original web interface for editing files, so I should be able to do it. Thanks for the update!

**Comment by MrHaroldA** [comment]
We are investigating if Geany would be a viable alternative for the huge Gitlab installation we have running now, and the only real blocker we are seeing now is the lack of conflict resolving withing the webinterface. This was mentioned in #14477 but lost traction and it never reached the [1.16](https://github.com/go-gitea/gitea/milestone/93) milestone.

@richmahn do you have any plans on implementing this? I have zero experience in coding for Gitea or Go, so I can't be of any help.

**Comment by richmahn** [comment]
@MrHaroldA Totally forgot I even mentioned I would work on this. Will put it on my todo list and see if I can take a stab at it in the coming week.

**Comment by Alexander-Zierhut** [comment]
@richmahn Still wish I had this feature sometimes for productivity sake 😆 By now, I had to learn git more in detail, can you imagine?

**Comment by ktpx** [comment]
I have a PR that now says 2 files are in conflic with the recieving repo., but i still see now way to resolve this...  How in earths name can I now fix this? Local repo says everything is fine. Won't let me delete the PR either.

**Comment by eUgEntOptIc44** [comment]
@richmahn any news on this?

**Comment by 6543** [comment]
@ktpx that's off topic ... but: merge target branch locally or rebase and push

**Comment by richmahn** [comment]
@eUgEntOptIc44 Sadly haven't had time to work on it, but have mentioned it to my team how badly it is needed. Hopefully can look at it soon, or if someone else wants to get this done, let me know. Should we do it exactly like Github does it, or have some extra functionality to making merging easier?

**Comment by eUgEntOptIc44** [comment]
@richmahn thank you for your update on this. I'd like to directly support on this by contributing code. Unfortunately my knowledge of go is very limited. However I hope that you or someone else might be able to do so.

**Comment by jayczech23** [comment]
@richmahn @eUgEntOptIc44 I am also interested in contributing to this feature that would be extremely helpful for my team's needs as well. 

I would also mention that the way Github handles merge conflict would be a great baseline for an initial implementation.

Please let me know if available for collaboration on this much needed feature.

**Comment by 6543** [comment]
I'm not aware of any draft so I would say go on and it is yours ;)
If you need help just ask.
for the UI side I would have a look if it can be done in vue ...
example:  #19650 ...

**Comment by richmahn** [comment]
@jayczech23 I'm still needing this and keep playing around with getting this implemented like Github. Can we collaborate together?

**Comment by richmahn** [comment]
@jayczech23 Are you on the Gitea Discord server? If we could touch base there I would be far more likely to get to work on this and better communication since I do everything with Discord. https://discord.gg/mPWThBne

**Comment by jayczech23** [comment]
@richmahn sounds great. I will sign up on the Gitea Discord and send you a message.
Thanks!

**Comment by Alexander-Zierhut** [comment]
Would love to get an occasional update on this or even help out. I am on the Gitea Discord as well and tagged you @richmahn

**Comment by richmahn** [comment]
@jayczech23 @Alexander-Zierhut Ok, now I'm more clear on you both wanting to help. Awesome. I won't let this drop now. Would be a cool hack-a-thon. My work pays me to work on Gitea, but usually working on our own fork of it and then I merge in from upstream (Gitea). I have many coworkers who use our copy of Gitea for editing text files (content) rather than code. 

A few things we need to figure out is first of all, I often get told when making a PR in Gitea that there is a conflict, due to different lines being edited, I think 2 apart, so the line between them still remains the same. This for some reason makes Gitea think there is a conflict. Yet if I merge master/main into the conflicting branch, it merges with no problem. That's one thing we need to solve....stop Gitea from declaring such a conflict (if it still is happening) as it is confusing.

Next is we need to mark up what a conflict editor UI should look like. Should it be just like Github, or something better?

**Comment by richmahn** [comment]
@jayczech23 @Alexander-Zierhut I have added you to a Github test repo so we can see how Github does it, mock up the Gitea, and implement all the different paths a user might take:

Accept the collaboration and see this PR:

https://github.com/richmahn/test/pull/3

**Comment by richmahn** [comment]
Here's a PR where the file was modified in base branch but deleted in the head branch. Github just says it is too complicated and says to deal with it on your computer. Wonder if there is a way we could deal with it in the UI? 

https://github.com/richmahn/test/pull/4

**Comment by richmahn** [comment]
Ditto with a binary file: https://github.com/richmahn/test/pull/5

**Comment by richmahn** [comment]
Large file with conflicts: https://github.com/richmahn/test/pull/6/conflicts

**Comment by richmahn** [comment]
If we make it exactly like Github we can then just copy their help pages:

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-using-the-command-line

(Maybe the last one Gitea already has?)

**Comment by richmahn** [comment]
So basically the base branch is being merged into the head branch to get the conflicts and the diff and then when the user submits the changes it updates with head branch with the base branch and fixes the conflicts. I wonder what the caveats are in doing this, such as as what if either branches are edited and pushed while the user is making their conflict resolution changes, if a temporary branch should be made to work from, etc. Thoughts? How do we properly map this out git-wise, gitea-wise and UI-wise? Do we lock the head branch or something? I wonder what Github does to handle conflict resolution and new changes while it is being done.

**Comment by xuan-nguyen-swe** [comment]
For simple conflicts, especially for writing documentation or books, this feature is really useful.
In these cases, I really don't want to wait until I can open my IDE to solve simple conflicts.

**Comment by NatashaCyn** [comment]
Any news with this issue?

**Comment by richmahn** [comment]
@NatashaCyn Funny, I just thought about the dire need my organization has for this feature, yet I still haven't gotten to it. I really would like there to be a 2-3 of us working on this...not that I couldn't program it, as I even made the file editor/committer for Gogs/Gitea interface (the original anyway, I'm sure it has been refactored), but to make sure we are doing it right, both how Github does, but also maybe in a way that could improve on that, would be great to have others working on it.

**Comment by richmahn** [comment]
@NatashaCyn I'm trying to get my company's fork and codebase up-to-date with Gitea 1.19.3 and release it. Once that is done, hopefully I can look more at the conflict resolver.

**Comment by richmahn** [comment]
Ok, getting my company fork updated from v1.20, lol. Pretty much done.

**Comment by ijustlovemath** [comment]
Just wanted to add that this feature would be fantastic! If implementation hasn't started yet, I'd like to say I love how clear the IntelliJ/PyCharm style of merge conflict resolution is:

![image](https://github.com/go-gitea/gitea/assets/12447024/83d67a93-ce37-4460-9479-397a21e8a674)

The sweet part about this implementation is two-fold:

1. Good labelling ("Your version", "Result", "Changes from <branch>")
2. The little >> button (next to line numbers) to say "I want this change from this version", which basically lets you build the exact set of changes that you want

Unfortunately not a Go coder, but definitely following along in case there's anything minor I can contribute!

**Comment by ijustlovemath** [comment]
The main challenge with any sort of conflict resolution UI will always be horizontal space. As long as there's a synchronized horizontal scrollbar and a minimum about of horizontal space (enough room for 90/150/180/240 columns + surrounding UI), it should be workable!

This type of interface probably leaves mobile users out to dry, but I wonder how many people are trying to do complex merge conflict resolution on their mobile devices to begin with

Another alternative could be to do it in stages, eg "select changes from your version", "select changes from remote version", "final review", which you could do in some kind of tabbed UI, but then you lose the clarity of having the full picture laid out before you.

Interested to hear any thoughts from others!

**Comment by puni9869** [comment]
I am excited to see this. @yardenshoham  thanks for accepting this.

**Comment by yardenshoham** [comment]
Just to clarify, I think this is a good idea. I don't plan on implementing it personally.

**Comment by richmahn** [comment]
@yardenshoham It's all yours! 

haha. I know that. I don't mind doing it myself. I just need to find time as it is on my comapny's time the work I put into Gitea.

**Comment by makar112233** [comment]
+1

**Comment by CelaniDe** [comment]
hello guys, any news on this?

**Comment by techknowlogick** [comment]
That was me ^
I'm testing the linked platform for bounties since bounty source went awol

**Comment by Bhavyajain21** [comment]
Can I be assigned?

**Comment by richmahn** [comment]
@oliverqx Want to team up? I've been meaning to get to this and my developers even asked about it recently and told them I would get to it once my big project that ended last quarter (March 31) was over, and now it is over. Interesting to see there is now a bounty on this.

**Comment by richmahn** [comment]
Not trying to step on any toes here, but seeing what happens if someone else also does an attempt.

**Comment by oliverqx** [comment]
you can go ahead, I think i'll be busy for the following couple of weeks

**Comment by richmahn** [comment]
Back to working on this finally. Just made a move across the country the last few months.

**Comment by HammyHavoc** [comment]
> Back to working on this finally. Just made a move across the country the last few months. 

Hope you've settled in well! :- ) Exciting stuff!

**Comment by HyCore** [comment]
hi everyone :D how can i increase the bounty haha, this feature would be dope to manage our conflicts in our gitea soc/siem/soar project.

**Comment by lunny** [comment]
Please visit here https://console.algora.io/org/CommitGo/bounties/community?fund=go-Gitea%2Fgitea%239014
